File: kotlin/luhn/src/main/kotlin/Luhn.kt

class Luhn {
    companion object {
        private fun doubleDigit(digit: Int): Int {
            return if (digit * 2 > 9) {
                digit * 2 - 9
            } else {
                digit * 2
            }
        }

        private fun doubleEvery2ndDigit(digit: Int, position: Int): Int {
            return if (position.rem(2) == 0) {
                digit
            } else {
                doubleDigit(digit)
            }
        }

        fun isValid(number: String): Boolean {
            val numbersOnly = number.filter(Character::isDigit).map(Character::getNumericValue)
            val invalidChars = number.filter { c: Char -> !(c.isDigit() || c.isWhitespace())}
            val doubledNumbers = numbersOnly.reversed().mapIndexed { index, digit -> doubleEvery2ndDigit(digit, index) }
            val sumOfNumbers = doubledNumbers.reduce { sum, i -> sum + i }

            return numbersOnly.size > 1 && invalidChars.isEmpty() && sumOfNumbers.rem(10) == 0

        }
    }
}

File: dart/luhn/lib/luhn.dart
import 'package:collection/collection.dart';

class Luhn {
  bool valid(String number) {
    if (number.trim().length <= 1) return false;
    if (RegExp(r'([^\ 0-9])').hasMatch(number)) return false;

    var digitsOnly = number.replaceAll(RegExp(r'[^0-9]'), "").split('');
    var processedDigits = digitsOnly.reversed.mapIndexed((i, digit) => _convertDigit(i, int.parse(digit)));
    var sumOfDigits = processedDigits.reduce((acc, digit) => acc + digit);

    return sumOfDigits % 10 == 0;
  }

  int _convertDigit(int i, int digit) {
    if (i % 2 == 0) {
      return digit;
    } else {
      digit = digit * 2;

      return digit > 9 ? digit - 9 : digit;
    }
  }
}
